### PR TITLE
feat: Replay Universal Filters

### PR DESCRIPTION
**Pull request by daibhin**
## Problem

Towards https://github.com/PostHog/posthog/issues/22054#issuecomment-2139714014

Since https://github.com/PostHog/posthog/pull/22707 introduced the new `UniversalFilters` component we can look to make the filtering experience in Replay more powerful

Relies on
https://github.com/PostHog/posthog/pull/22625
https://github.com/PostHog/posthog/pull/22707

## Changes

filters

- Add `UniversalFilters` and nested groups UI
- Add feature flag for universal filtering in Replay
- Write a conversion method to convert universal filters to existing / legacy filters
   - HogQL filtering still isn't released on the backend
   - Changing the backend would require supporting two code paths
   - Given one of them will be gone shortly it felt easier to convert universal filters to legacy rather than the other way around
   - Before we ever release this to customers / beta test we should make sure universal is the default (and legacy filters are converted forward)

Coming in followup PRs:
- [ ] Top level duration filter
- [ ] Inline console logs filter
- [ ] Custom operand
- [ ] Nested groups

### CODE REVIEW

**Comment by daibhin** [comment]
@pauldambra just a heads up so you get a sense of the HogQL UI work. I'll probably keep contributing to this branch until tomorrow at least given everything will be shipped behind a flag. Realised it would be easier (and probably safer) to convert the new filtering schema to the existing one on the frontend so there should be no backend changes needed to get this out :)

File: frontend/__snapshots__/filters-universalfilters--default--dark.png
Comment (pauldambra) [inline frontend/__snapshots__/filters-universalfilters--default--dark.png:1]: (commenting here so we can thread)
Screenshot 2024-06-04 at 09 53 01
Screenshot 2024-06-04 at 09 53 23

i worry about the discoverability of the extra filter button inside the taxonomic property pill here... but then I think the same on the insights page

ultimately they should probably be the same but that could follow

**Review by pauldambra — COMMENTED** [review]
this is really nice... nicer than i expected
we'd need to find a way to tell existing users about it, or use feature preview to opt in/out for a little while maybe
but i really want to remove the custom UI-iness of replay

File: frontend/__snapshots__/filters-universalfilters--default--dark.png
Comment (daibhin) [reply frontend/__snapshots__/filters-universalfilters--default--dark.png:1]: Yep, certainly open to aligning on something that works for all filters across the app. I'm guessing @corywatilo has some opinions on that.

The latest commit ships the concept of "headless" filters, decoupling the UI from the logics. It's down to the implementor (session replay being the first) to decide on the UI while getting all the benefit of the more complex logic code / structure.
I feel strongly that we should keep UI & logic distinction even if we find a UI that works for every case (which I suspect is preferable from a UX point of view)

File: frontend/src/scenes/session-recordings/filters/RecordingsUniversalFilters.tsx [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+                                    values: universalFilters.filter_group.values,
+                                },
+                            })
+                        }}
+                        disabledReason="'Or' filtering is not supported yet"
+                        topLevelFilter={true}
+                        suffix={['filter', 'filters']}
+                    />
+                </div>
+            </div>
```
Comment (daibhin) [inline frontend/src/scenes/session-recordings/filters/RecordingsUniversalFilters.tsx:74]: Everything up to this point is just copied from our advanced filters

File: frontend/src/scenes/session-recordings/filters/RecordingsUniversalFilters.tsx [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+                    ]}
+                    onChange={(filterGroup) => {
+                        setUniversalFilters({
+                            ...universalFilters,
+                            filter_group: filterGroup,
+                        })
+                    }}
+                >
+                    <RecordingsUniversalFilterGroup />
+                    <UniversalFilters.AddFilterButton />
```
Comment (daibhin) [inline frontend/src/scenes/session-recordings/filters/RecordingsUniversalFilters.tsx:96]: Only allow new filters at the root level

File: frontend/src/scenes/session-recordings/filters/RecordingsUniversalFilters.tsx [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+const RecordingsUniversalFilterGroup = (): JSX.Element => {
+    const { filterGroup } = useValues(universalFiltersLogic)
+    const { replaceGroupValue, removeGroupValue } = useActions(universalFiltersLogic)
+
+    return (
+        <>
+            {filterGroup.values.map((filterOrGroup, index) => {
+                return isUniversalGroupFilterLike(filterOrGroup) ? (
+                    <UniversalFilters.Group key={index} index={index} group={filterOrGroup}>
+                        <RecordingsUniversalFilterGroup />
```
Comment (daibhin) [inline frontend/src/scenes/session-recordings/filters/RecordingsUniversalFilters.tsx:112]: Allows recursion to support multiple filter groups (upcoming PR)

File: frontend/src/types.ts [hunk 4]
```diff
@@ -972,6 +973,17 @@ export interface RecordingFilters {
     filter_test_accounts?: boolean
 }
 
+export interface RecordingUniversalFilters {
+    /**
+     * live mode is front end only, sets date_from and date_to to the last hour
+     */
+    live_mode?: boolean
```
Comment (pauldambra) [inline frontend/src/types.ts:980]: aha... I need to unwind the artificial lag changes 🙈

**Review by pauldambra — APPROVED** [review]
LGTM behind a flag and things work well with the flag off...

the existing filter UI doesn't get hidden when I turn the new one on... i guess it should?

Screenshot 2024-06-10 at 09 41 17

**Pull request merged**